### DISCUSSION TITLE
Automatic build infrastructure

### DISCUSSION THREAD

**Discussion by valinet** [valinet/ExplorerPatcher#276] [Announcements]
Hi everyone,

Recently, some people (#264, #265, #267) encountered a few problems regarding the DLL suddenly becoming flagged by Windows Defender as malicious. To address this situation, I have set up the infrastructure to have the builds be automatically generated on GitHub's infrastructure. Every time a new commit is made, the latest binaries will be compiled on GitHub's infrastructure and a new release with them will be generated and made available. It's not a direct fix for the situation described, but rather a way to ensure the uploaded builds are as trusted, genuine and untouched as possible. 

Also, the added benefit is that releases now always are up to date, in sync with the main repository, since I do not have to publish them manually anymore. No more need for sharing various development versions in different threads or waiting for the latest version to release. You now have access to the most up to date version at all times, without having to set up your own build infrastructure, with all the benefits and downsides this implies. The latest release will always be generated from the most recent code available in the repository. For the moment, I decided to publish them as actual releases, not pre-releases, as I think the model should work fine.

P.S. I have published the latest pre-release as a full release as well, 22000.282.32.6, the only changes to .5 are just a version bump to coincide with the introduction of this automatic build infrastructure.

Thank you,
@valinet

**Comment by Speedy37** [comment]
You can have a look at https://github.com/Speedy37/ExplorerPatcher/pull/1 if you just want to build artifacts without creating a release.
Artifacts are then available here: https://github.com/Speedy37/ExplorerPatcher/actions/runs/1424284475
I would only generate release build for the master branch, otherwise, pull requests won't be ✔️.

**Reply by valinet** [reply]
I want to have it generate releases for stuff in master so that people have access to the latest versions without me needing to manually upload, plus so that the binaries come from an untouched source so to say.

So basically what you're saying is to integrate [this](https://github.com/Speedy37/ExplorerPatcher/pull/1/commits/977a1b2419ffd8218bb5618757a2a308978eabc8) as well so that it also builds it for any push on any branch, right?

**Reply by Speedy37** [reply]
I would suggest building artifacts for any branch (so it can be tested) and limit the auto release process on master only.
This way you can be sure any pull request will build successfully before merging.

This part:
```yml
    - name: Upload artifacts
      uses: actions/upload-artifact@v2
      with:
        name: bin
        path: |
          build/Release/
```

**Reply by valinet** [reply]
What's that part? Exactly what to change in this? https://github.com/valinet/ExplorerPatcher/blob/master/.github/workflows/build.yml

**Reply by Speedy37** [reply]
I'll do a pull request.

**Reply by valinet** [reply]
You can post here as well, I have the file open in my editor anyway......

**Reply by Speedy37** [reply]
See https://github.com/valinet/ExplorerPatcher/pull/277, it worked on my fork.

**Reply by valinet** [reply]
Thanks. I will merge it.

**Reply by Speedy37** [reply]
I forgot `on: [push, pull_request]` in my initial commit, pull request is now uptodate.

**Reply by valinet** [reply]
Yeah, no worries, I managed to see it in time :D Thanks

**Reply by valinet** [reply]
That's the issue I had previously as well, that's why I restricted it to `master` only, this also runs for the tags that are generated when I upload a release. It's not that big of an issue, but it is useless, do you know if we can turn it off for those? Maybe I can prefix the automatic release tags with some characters and identify based on that...?

**Reply by valinet** [reply]
I mean, yeah, I can add some if's like those that you added. Is there something more intelligent so to speak or that's fine?

**Reply by valinet** [reply]
Nvm, solved, `if`s work for whole jobs as well, apparently.

**Reply by valinet** [reply]
Nvm x2, the correct way to do it is by adjusting the `on` trigger, done that eventually.
